### PR TITLE
Fix spurious directory separators in `mixer /listmidi` output

### PR DESCRIPTION
**Pull request by dreamer**
Quote from commit message:

Command 'mixer /listmidi' depends on OS env variables and user settings
to look for .sf2 installed on the system. This means that in some cases
we had a problem, e.g. when running:

    XDG_DATA_DIRS=/usr/share dosbox

command was printing (correctly):

    * default.sf2 - /usr/share/soundfonts/default.sf2

but for:

    XDG_DATA_DIRS=/usr/share/ dosbox

command was printing spurious directory separators:

    * default.sf2 - /usr/share//soundfonts/default.sf2

This is annoying, because to support accidental lack of separator
implemenation was sometimes adding spurious separators already, so the
problem was *always* visible when system-wide soundfont was installed.

Use std::filesystem ability to join path elements in cross-platform way
to prevent this problem and avoid various edge-cases.

### CODE REVIEW

**Comment by kcgen** [comment]
Code looks great, @dreamer - especially with `std::fs` sf2 objects being created once at the initial points, and then carried throughout the programs lifetime (instead of using std::string and using intermediate std::fs objects).
Will test on my side too.

File: src/midi/midi_fluidsynth.cpp [hunk 1]
```diff
@@ -134,39 +134,39 @@ std::tuple<std::string, int> parse_sf_pref(const std::string &line,
...
 #elif defined(MACOSX)
 
-static std::deque<std::string> get_data_dirs()
+static std::deque<std_fs::path> get_data_dirs()
 {
 	return {
-	        CROSS_GetPlatformConfigDir() + "soundfonts/",
-	        CROSS_ResolveHome("~/Library/Audio/Sounds/Banks/"),
+	        std_fs::path(CROSS_GetPlatformConfigDir()) / "soundfonts",
+	        std_fs::path(CROSS_ResolveHome("~/Library/Audio/Sounds/Banks")),
```
Comment (kcgen) [inline src/midi/midi_fluidsynth.cpp:151]: Looking at CROSS_GetPlatformConfigDir(), maybe it can cache and return a std::fs::path object (to avoid repeated wrapper construction from strings)?

It's not used in much code, so ripples should be pretty minor: 

```
git grep  -l CROSS_GetPlatformConfigDir

gui/sdlmain.cpp
midi/midi_fluidsynth.cpp                             midi/midi_mt32.cpp
misc/cross.cpp                                       misc/support.cpp
```

**Review by kcgen — APPROVED** [review]
One minor comment, if you think it's doable.

File: src/midi/midi_fluidsynth.cpp (hunk 1 above)
Comment (dreamer) [reply src/midi/midi_fluidsynth.cpp:151]: I was wondering if I should just clean it up as well :smile:. I'll try to do it, and maybe even limit the usage of `Cross::GetPlatformConfigDir` at the same time.

Comment (kcgen) [reply src/midi/midi_fluidsynth.cpp:151]: Very nice :-)
Approved either approach; please feel free to merge when passing!

Comment (dreamer) [reply src/midi/midi_fluidsynth.cpp:151]: Of course the refactor turned out to be non-trivial... Decorated code with asserts and deprecations until it's done (which means CI will fail on deprecations for now). Setting PR to draft for now.

File: src/midi/midi_fluidsynth.cpp [hunk 2]
```diff
@@ -134,39 +134,39 @@ std::tuple<std::string, int> parse_sf_pref(const std::string &line,
 
 #if defined(WIN32)
 
-static std::deque<std::string> get_data_dirs()
+static std::deque<std_fs::path> get_data_dirs()
 {
 	return {
-	        CROSS_GetPlatformConfigDir() + "soundfonts\\",
+	        get_platform_config_dir / "soundfonts",
```
Comment (dreamer) [inline src/midi/midi_fluidsynth.cpp:140]: Ugh... Why this compiles?!

**Comment by johnnovak** [comment]
Yeah, common problem... I always do a `path.stripSuffix('/')` on every directory type path I read from a config at work before I mangle it further. But path handling functions are the best, agreed.

File: src/midi/midi_fluidsynth.cpp (hunk 2 above)
Comment (johnnovak) [reply src/midi/midi_fluidsynth.cpp:140]: Btw, is that `C:\soundfonts` path needed? I'm bothered by it because if it's supported I need to document it in the manual, and such silly hardcoded paths make us look like amateurs...

I guess if it's just some weird legacy backward compatibility thing we need for eXoDOS or something then can keep it and perhaphs don't even menion it in the manual.

File: src/midi/midi_fluidsynth.cpp (hunk 1 above)
Comment (johnnovak) [reply src/midi/midi_fluidsynth.cpp:151]: Generally, I don't care optimising code than only gets called a handful of times and optimise for simplicity and readability. Why does it matter that these objects are not cached We're talking about less than 10 calls perhaps when running the program, not millions.

**Review by johnnovak — APPROVED** [review]
Looking good @dreamer; just some minor quibblings from me 😄

File: src/midi/midi_fluidsynth.cpp (hunk 2 above)
Comment (dreamer) [reply src/midi/midi_fluidsynth.cpp:140]: I added this path because of: https://www.fluidsynth.org/api/fluidsettings.xml#synth.default-soundfont

On Windows there's not a single good place to look up systemwide-installed soundfonts :disappointed:  I decided that following FluidSynth gives us the best chance of actually finding something (perhaps I should've left a comment).

File: src/midi/midi_fluidsynth.cpp (hunk 1 above)
Comment (kcgen) [reply src/midi/midi_fluidsynth.cpp:151]: @johnnovak , fully agree that caching isn't needed.

Comment (dreamer) [reply src/midi/midi_fluidsynth.cpp:151]: I don't care about optimisation in this case either (it's just a happy accident) - but I do care about deprecating/removing functions that return paths as strings (for code readability).

Old DOSBox code has this annoying  property that it (usually, not always) expects directory separator to be returned from functions returning directory paths... Which leads to confusing code, as readers can't easily distinguish if code is joining paths or appending suffix to path.

File: src/midi/midi_fluidsynth.cpp (hunk 2 above)
Comment (dreamer) [reply src/midi/midi_fluidsynth.cpp]: Oh, @johnnovak I wouldn't mind if we skipped documenting it ;) Simply if user has FluidSynth installed then we might find some fonts here.

Timidity does not have default directory for sf2 files, MuseScore is missing documentation in this regard...

Comment (johnnovak) [reply src/midi/midi_fluidsynth.cpp]: > I added this path because of: https://www.fluidsynth.org/api/fluidsettings.xml#synth.default-soundfont
> 
> On Windows there's not a single good place to look up systemwide-installed soundfonts 😞 I decided that following FluidSynth gives us the best chance of actually finding something (perhaps I should've left a comment).

Yeah, I think it's a bit of a moot point, whether FluidSynth uses that path or not. I think the normal use case is that DOSBox has a SoundFont folder, the user copies the .sf2 files in there, job done.

In reality, nothing uses SF2 files much these days... I'm also a hobby musician, and I haven't cared about SoundFonts since about 2002 or so, there are so much better options. As I see it, it's a niche format pretty much only used by a handful of emulators and maybe a few MIDI players. I don't think anyone wants to install "system wide SoundFonts" as MIDI is only used by emulators such as DOSBOx, and even for hobbyist music production stuff, there are vastly better options... Hell, there's not even a *single* good SF2 player in VSTi format. Sure, many samplers such as Native Instruments Kontakt can import SF2 files... but there's not much point as there are so many better sample libraries.

Maybe the situation is different on Linux, I don't know...

TL;DR there's not really such a thing on Windows as "system wide SoundFonts", because apart from people running emulators, virtually no one cares about MIDI, and even people who do don't care about SoundFonts as it's an old legacy technology supersede like 2 decades ago. Yeah, good for General MIDI emulation, but not much else, to be honest.

Comment (johnnovak) [reply src/midi/midi_fluidsynth.cpp]: > MuseScore is missing documentation in this regard...

It's here:
https://musescore.org/en/handbook/3/soundfonts-and-sfz-files

Default locations (the user can copy SF2 files here):
Windows: %HOMEPATH%\Documents\MuseScore3\Soundfonts

System locations:
Windows x86 (32-bit) / MuseScore x86: %ProgramFiles%\MuseScore 3\sound\MuseScore_General.sf3
Windows x64 (64-bit) / MuseScore x86: %ProgramFiles(x86)%\MuseScore 3\sound\MuseScore_General.sf3
Windows x64 (64-bit) / MuseScore x86_64: %ProgramFiles%\MuseScore 3\sound\MuseScore_General.sf3

So as I said, every app has its own SoundFont directory, that's pretty standard. Personally, if I really wanted to reuse SF2 files, I'd save them in somewhere such as D:\Music\Data\SF2 or something, then point the apps to that folder in their preferences.

Comment (dreamer) [reply src/midi/midi_fluidsynth.cpp]: > Maybe the situation is different on Linux, I don't know... Yeah, I guess there are "free sample libraries" available in SF2, but honestly, all the good stuff is paid... No musician cares, except for some die-hard Linux hobbyists perhaps.

On Linux and macOS the soundfonts can be installed system-wide, and it is normal feature, so Windows is odd-one out in this regard.

On Linux in particular, we rely on that - during package installation some soundfont can be automatically installed as a dependency, so users litearally can just switch to `mididevice = fluidsynth` without ever downloading any soundfonts manually. At least that's how I packaged it for Fedora and encouraged other packagers to do the same - it just works out of the box for users 😄.

On macOS, the well-known place for installation of soundfonts is shared between all applications, and additionally tied to CoreMidi system interface (or CoreAudio, I don't remember) system interface (so it actually affects users who set `mididevice = coremidi`).

For Windows I would prefer to keep the directory in list (just in case some user installs or builds FluidSynth on their machine), but I see **no reason for documenting it** - we can just keep it as a hidden nicety feature. I will add comment in code to explain.

File: src/midi/midi_fluidsynth.cpp (hunk 1 above)
Comment (dreamer) [reply src/midi/midi_fluidsynth.cpp:154]: Ugh, this task grew a bit due to all places I needed to touch to remove old `CROSS_GetPlatformConfigDir`; I think reviewing commit-by-commit might be easier.

@kcgen I will submit changes for `CROSS_ResolveHome` in another PR, I think this one is big enough. Ready for re-review 😄.

File: src/midi/midi_fluidsynth.cpp (hunk 2 above)
Comment (johnnovak) [reply src/midi/midi_fluidsynth.cpp]: > For Windows I would prefer to keep the directory in list (just in case some user installs or builds FluidSynth on their machine), but I see **no reason for documenting it** - we can just keep it as a hidden nicety feature. I will add comment in code to explain.

Yeah that's the best then.

I'm aware this modular, system-wide approach is much more common on macOS and Linux, it's just on Windows only a handful of things work like that (VST plugins are system-wide, for example; but with general DLL files it's a mixed bag...)

**Comment by kcgen** [comment]
Tested all the commits (the nice one-by-one replacement sequence will also help bisect any future issues), and all are working. 
All comments addressed , so merging.
